exonum/src/runtime/execution_context.rs

```diff
@@ -0,0 +1,409 @@
+// Copyright 2020 The Exonum Team
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use crate::{
+    blockchain::Schema as CoreSchema,
+    crypto::{Hash, PublicKey},
+    helpers::{Height, ValidateInput},
+    merkledb::{access::Prefixed, BinaryValue, Fork},
+    runtime::{
+        ArtifactId, BlockchainData, CallSite, CallType, Caller, CoreError, Dispatcher,
+        DispatcherSchema, ExecutionError, InstanceDescriptor, InstanceId, InstanceQuery,
+        InstanceSpec, InstanceStatus, MethodId, SUPERVISOR_INSTANCE_ID,
+    },
+};
+
+/// Provides the current state of the blockchain and the caller information for the call
+/// which is being executed.
+///
+/// The call can mean a transaction call, `before_transactions` / `after_transactions` hook,
+/// or the service constructor invocation.
+#[derive(Debug)]
+pub struct ExecutionContext<'a> {
+    /// The current state of the blockchain. It includes the new, not-yet-committed, changes to
+    /// the database made by the previous transactions already executed in this block.
+    pub fork: &'a mut Fork,
```

Comment:
We have a major problem here: having `&mut Fork` publicly available to services negates all the benefits from having data separation via `BlockchainData`! I know it may be frustrating, but I think the best way to deal with this is to leave a separate `CallContext` type in the Rust runtime crate as a thin wrapper around `ExecutionContext`, copying methods that we want to expose to the Rust services.

Reply:
Ouch. In theory, we can create type-based restriction for the `Fork` access (see [playground example](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=b3b27efcc2e5bf65cc82133e9d2cd38a) ), expecting the runtimes to convert context into unprivileged, but not sure if such a complication worth it.

Reply:
Generic approach with a separate type is more flexible, so IMO we should go with it.

Reply:
Sure.